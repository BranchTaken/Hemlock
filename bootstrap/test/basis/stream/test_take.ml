open! Basis.Rudiments
open! Basis
open Stream
open Format

let test () =
  let ppt = (pp Uns.pp) in
  printf "@[<h>";
  let rec test_take_up_to i l n =
    (* l + 1 so that we test taking one more than the stream contains *)
    match i <= l + 1L, l <= n with
    | _, false -> ()
    | false, _ -> test_take_up_to 0L (succ l) n
    | true, true -> begin
        let t = init (0L =:< l) ~f:(fun i -> i) in
        let t' = take i t in
        printf "take %a %a = %a\n" ppt t Uns.pp i ppt t';
        test_take_up_to (succ i) l n
      end in
  test_take_up_to 0L 0L 3L;
  printf "@]"

let _ = test ()
