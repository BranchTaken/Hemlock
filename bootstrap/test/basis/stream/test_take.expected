take Nil 0 = Nil
take Nil 1 = Nil
take (0 Nil) 0 = Nil
take (0 Nil) 1 = (0 Nil)
take (0 Nil) 2 = (0 Nil)
take (0 (1 Nil)) 0 = Nil
take (0 (1 Nil)) 1 = (0 Nil)
take (0 (1 Nil)) 2 = (0 (1 Nil))
take (0 (1 Nil)) 3 = (0 (1 Nil))
take (0 (1 (2 Nil))) 0 = Nil
take (0 (1 (2 Nil))) 1 = (0 Nil)
take (0 (1 (2 Nil))) 2 = (0 (1 Nil))
take (0 (1 (2 Nil))) 3 = (0 (1 (2 Nil)))
take (0 (1 (2 Nil))) 4 = (0 (1 (2 Nil)))
